Implement 'omitemptyrecursive' tag

akalin-keybase opened the pull request:
This is in spirit a reimplementation of a previous PR that implemented an
'omitemptycheckstruct' tag. This is necessary because, even though we
already have a RecursiveEmptyCheck field in EncoderOption, flipping it
on results in undesired behavior, like existing pointer fields with omitempty
being checked recursively, and not being able to specify otherwise.

This new 'omitemptyrecursive' tag is a combination of 'omitempty' and a
per-field version of RecursiveEmptyCheck.

akalin-keybase committed:
Add test for omitemptyrecursive (to be added)

akalin-keybase committed:
Implement 'omitemptyrecursive' keyword

akalin-keybase commented:
Previous PR is #4.

strib reviewed (approved):
> This is necessary because, even though we already have a RecursiveEmptyCheck field in EncoderOption, flipping it on results in non-obvious behavior, like pointer fields with omitempty doing recursive checks.

The code looks fine, but I don't really understand this justification.  Why wouldn't you expect the recursive empty check to descend into pointers?  The documentation for `RecursiveEmptyCheck` says explicitly it will do this.

Is this more about having stricter control over the performance setbacks of `RecursiveEmptyCheck`, to confine it only to fields we care about?

Can you please update the `RecursiveEmptyCheck` documentation string to mention this tag and explain why you'd want to use it instead of the global flag?

akalin-keybase committed:
Add documentation for 'omitemptyrecursive' tag

akalin-keybase commented:
Yeah, I guess I was conditioned by my earlier implementation, which flipped on a flag to descend into structs but not pointers. I added documentation for the tag, and fixed the PR description.

It's not really about performance setbacks, but about control, and the desire to not disrupt anything when updating codec, since turning on the global flag would change behavior, like for pointers fields with omitempty, for which there is no way to preserve the existing shallow compare behavior with the global flag on, or for structs incorrectly tagged with omitempty.
